### PR TITLE
#159204716 Completes descriptive validation errors for user registration

### PR DESCRIPTION
**Pull request by chukwuemekachm**
#### What does this PR do?
- Completes descriptive validation errors for user registration
#### Description of Task to be completed?
- Write unit tests for endpoint
- Write tests for base url and un-available urls
- Install validatorjs
- Modify ./server/index.js by removing sessions and changing the message on the base url
- Add UserValidation.js to handle all user validations
- Add .sequelizerc to enable sequelize-cli operations from the commandline
- Modify ./server/models/user.js with user properties
- Add user migration file
- Add ./server/routes/auth.js to handle authentication requests
- Add signupValidation to implement user signup validation
- Add AuthController to handle all user authentication
- Add signUpUser to to implement user signup
- Test endpoint using POSTMAN
- Update .env-sample
#### How should this be manually tested?
- Clone the repository
- Start up the command line /Terminal
- Change to the repo directory
- Run `npm install`
- Run `npm run migrate`
- Run `npm run dev`
- Use POSTMAN to create a POST request to `/api/auth/signup`
#### Any background context you want to provide?
- Create a Postrgresql database before you start
- Request payload for the POST request

``` 
{
"firstName": "John",
"lastName": "Doe",
"email": "johndoe@gmail.com",
"username": "johnny",
"password": "Twilight2",
"confirmPassword": "Twilight2"
}
```

#### What are the relevant pivotal tracker stories?
[#159204716](https://www.pivotaltracker.com/story/show/159204716)
#### Screenshots (if appropriate)

screen shot 2018-08-01 at 8 37 34 am

screen shot 2018-08-01 at 8 39 35 am

screen shot 2018-08-01 at 8 40 05 am

#### Questions:
- None

### CODE REVIEW

**Comment by yomigeek** [comment]
I must say you did a very good job. I think it would be much better if the validation files are placed within a validation folder in the middlewares folder because this will ensure adequate files structure and allows for easier identification of each file purpose.

File: server/middlewares/UserValidation.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+class UserValidation {
+  /**
+   * @description Validates the request payload to create a user
+   * @param {Object} req The request object
+   * @param {Object} res The response object
+   * @param {Object} next The next middleware
+   * @returns The next middleware to handle the user signup
+   */
+  static singupValidation(req, res, next) {
```
Comment (yomigeek) [inline server/middlewares/UserValidation.js]: NOTE: The method name should be 'signup' instead of 'singup'.

**Review by Oluwafayokemi — CHANGES_REQUESTED** [review]
Hi Chima! Good job. 
But I noticed you removed the passport session in the root file, doing that could cause an error when implementing social login because the session is important for the passport.js to run. Please, can you return the passport session script back to the root folder so as to prevent any form of conflict for other tasks that would need it?

**Review by Veraclins — CHANGES_REQUESTED** [review]
@chukwuemekachm Nice job bro. This is really beautiful.
Just a few recommendations
1. For consistency with declaring a function before using them, I suggest you take the declaration of stripUser and userExists.
2. Use only one status code for validation errors.

File: server/tests/1userSignup.test.js [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+  });
+
+  it('should return 409 when a user with the username already exists', (done) => {
+    chai.request(app).post('/api/auth/signup').send({
+      firstName: 'John',
+      lastName: 'Doe',
+      email: 'johndoe1@gmail.com',
+      username: 'Johnny',
+      password: 'Opcut2way',
+      confirmPassword: 'Opcut2way',
```
Comment (tonymontaro) [inline server/tests/1userSignup.test.js]: These are some excellent tests! Good job completing the task and testing it extensively. We should try as much as possible to write DRY code. You may not be sending the same object to the '/api/auth/signup' route, but they do have a lot in common. Is there a way to minimize this duplication?

Comment (tonymontaro) [reply server/tests/1userSignup.test.js]: Do remember to update the .eslintrc.json to fix un-necessary hound comments.

**Review by Oluwafayokemi — APPROVED** [review]

File: server/index.js [hunk 3]
```diff
@@ -21,9 +20,6 @@ app.use(morgan('dev'));
 app.use(bodyParser.urlencoded({ extended: false }));
 app.use(bodyParser.json());
 
-app.use(methodOverride());
-app.use(express.static(`${__dirname}/public`));
-
 app.use(
   session({
     secret: 'authorshaven',
```
Comment (Oluwafayokemi) [inline server/index.js]: Chima! Please return the whole session script as it was. It was already set to false

**Review by Oluwafayokemi — CHANGES_REQUESTED** [review]
Please return session script as it was, thank you

File: server/index.js (hunk 3 above)
Comment (chukwuemekachm) [reply server/index.js]: Thanks for your observation, but that is not a session script. @Oluwafayokemi. And the folder you are referring to don't exist.

**Review by Veraclins — APPROVED** [review]

File: server/middlewares/validations/UserValidation.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+      confirmPassword: 'required|alpha_num|min:8|max:20|same:password',
+    };
+
+    const validator = new Validator(req.body, userProperties);
+    validator.passes(() => {
+      if (UserValidation.passwordCheck(req.body.password)) return next();
+      return res.status(400).json({
+        status: 'error',
+        errors: {
+          password: ['Password most contain an Upper case letter, a lower case letter and a number.'],
```
Comment (yomigeek) [inline server/middlewares/validations/UserValidation.js]: Hi Chima, I see you have made changes to some of the vocabulary errors, however kindly change the 'most' here to 'must'.
Thanks.

**Review by madeofhuman — APPROVED** [review]

**Review by Oluwafayokemi — COMMENTED** [review]
Oh! it is not a script! can you tell me what it is called so that I will know for the next time? Thank you.
Nice job! Weldone

**Review by Oluwafayokemi — APPROVED** [review]

**Review by yomigeek — APPROVED** [review]
good!

**Pull request merged**